### PR TITLE
feat(chart): Adding namespace override feature

### PR DESCRIPTION
**Pull request by rupeshnemade**
### Checklist

- [x] Changelog updated or skip changelog label added
- [x] Documentation updated
- [x] Template tests added for new features
- [x] Integration tests added or modified for major features

### CODE REVIEW

**Review by swiatekm-sumo — COMMENTED** [review]
If you want to add this feature, you should add it for all the resources, not just logs. It should be a top-level `namespaceOverride` key.

**Comment by rupeshnemade** [comment]
@swiatekm-sumo Namespace override is added at top-level. Please review.

File: deploy/helm/sumologic/templates/_helpers/_common.tpl [hunk 1]
```diff
@@ -18,6 +18,17 @@ We truncate at 63 chars because some Kubernetes name fields are limited to this
 {{- end -}}
 {{- end -}}
 
+{{/*
+Allow the release namespace to be overridden
+*/}}
+{{- define "sumologic.namespace" -}}
+  {{- if .Values.namespaceOverride -}}
+    {{- Values.namespaceOverride -}}
```
Comment (swiatekm-sumo) [inline deploy/helm/sumologic/templates/_helpers/_common.tpl:26]:
```suggestion
    {{- .Values.namespaceOverride -}}
```

**Review by swiatekm-sumo — COMMENTED** [review]
Can you add the following test to tests/helm/common_test.go:

```go
func TestNamespaceOverride(t *testing.T) {
	valuesFilePath := path.Join(testDataDirectory, "everything-enabled.yaml")
	namespaceOverride := "override"
	renderedYamlString := RenderTemplate(
		t,
		&helm.Options{
			ValuesFiles: []string{valuesFilePath},
			SetStrValues: map[string]string{
				"sumologic.accessId":  "accessId",
				"sumologic.accessKey": "accessKey",
				"namaespaceOverride":  namespaceOverride,
			},
			Logger: logger.Discard, // the log output is noisy and doesn't help much
		},
		chartDirectory,
		releaseName,
		[]string{},
		true,
		"--namespace",
		defaultNamespace,
	)

	// split the rendered Yaml into individual documents and unmarshal them into K8s objects
	// we could use the yaml decoder directly, but we'd have to implement our own unmarshaling logic then
	renderedObjects := UnmarshalMultipleFromYaml[unstructured.Unstructured](t, renderedYamlString)

	for _, renderedObject := range renderedObjects {
		if !isSubchartObject(&renderedObject) && renderedObject.GetKind() != "List" {
			object := renderedObject
			objectName := fmt.Sprintf("%s/%s", object.GetKind(), object.GetName())
			t.Run(objectName, func(t *testing.T) {
				require.Equal(t, namespaceOverride, object.GetNamespace())
			})
		}
	}
}
```

This will catch the remaining places where you need to add the namespace block.

File: deploy/helm/sumologic/templates/_helpers/_common.tpl (hunk 1 above)
Comment (rupeshnemade) [reply deploy/helm/sumologic/templates/_helpers/_common.tpl]: Changes done as expected.

**Comment by rupeshnemade** [comment]
> Can you add the following test to tests/helm/common_test.go:
> 
> ```go
> func TestNamespaceOverride(t *testing.T) {
> 	valuesFilePath := path.Join(testDataDirectory, "everything-enabled.yaml")
> 	namespaceOverride := "override"
> 	renderedYamlString := RenderTemplate(
> 		t,
> 		&helm.Options{
> 			ValuesFiles: []string{valuesFilePath},
> 			SetStrValues: map[string]string{
> 				"sumologic.accessId":  "accessId",
> 				"sumologic.accessKey": "accessKey",
> 				"namaespaceOverride":  namespaceOverride,
> 			},
> 			Logger: logger.Discard, // the log output is noisy and doesn't help much
> 		},
> 		chartDirectory,
> 		releaseName,
> 		[]string{},
> 		true,
> 		"--namespace",
> 		defaultNamespace,
> 	)
> 
> 	// split the rendered Yaml into individual documents and unmarshal them into K8s objects
> 	// we could use the yaml decoder directly, but we'd have to implement our own unmarshaling logic then
> 	renderedObjects := UnmarshalMultipleFromYaml[unstructured.Unstructured](t, renderedYamlString)
> 
> 	for _, renderedObject := range renderedObjects {
> 		if !isSubchartObject(&renderedObject) && renderedObject.GetKind() != "List" {
> 			object := renderedObject
> 			objectName := fmt.Sprintf("%s/%s", object.GetKind(), object.GetName())
> 			t.Run(objectName, func(t *testing.T) {
> 				require.Equal(t, namespaceOverride, object.GetNamespace())
> 			})
> 		}
> 	}
> }
> ```
> 
> This will catch the remaining places where you need to add the namespace block.

Changes done as expected.

**Comment by swiatekm-sumo** [comment]
Allright, now please make CI happy and we should be good to go.

**Comment by rupeshnemade** [comment]
@swiatekm-sumo Trying to resolve unite test failure but looks like we are using hard-coded 'sumologic' namespace whereas we can't provide fix values in values.yaml.
image

Any suggestions?

**Comment by swiatekm-sumo** [comment]
> @swiatekm-sumo Trying to resolve unite test failure but looks like we are using hard-coded 'sumologic' namespace whereas we can't provide fix values in values.yaml. image
> 
> Any suggestions?

I think running `make regenerate-goldenfiles` will fix these kinds of errors for you.

File: deploy/helm/sumologic/templates/logs/collector/otelcol/daemonset.yaml [hunk 2]
```diff
@@ -13,7 +13,8 @@ apiVersion: apps/v1
 kind: DaemonSet
 metadata:
   name: {{ printf "%s%s" (include "sumologic.metadata.name.logs.collector.daemonset" $ctx) $instance | trunc 63 | trimSuffix "-" }}
-{{- if $.Values.otellogs.daemonset.annotations }}
+  namespace: {{ template "sumologic.namespace" . }}
```
Comment (swiatekm-sumo) [inline deploy/helm/sumologic/templates/logs/collector/otelcol/daemonset.yaml:16]: This happens in a loop with a context different than the default. We need to explicitly pass the global context like so:
```suggestion
  namespace: {{ template "sumologic.namespace" $ctx }}
```

File: tests/helm/common_test.go [hunk 3]
```diff
@@ -242,3 +242,40 @@ func GetPodSpec(object unstructured.Unstructured) (*corev1.PodSpec, error) {
 		return nil, nil
 	}
 }
+
+func TestNamespaceOverride(t *testing.T) {
+	valuesFilePath := path.Join(testDataDirectory, "everything-enabled.yaml")
+	namespaceOverride := "sumologic"
```
Comment (swiatekm-sumo) [inline tests/helm/common_test.go:248]: Please set this to a value other than `sumologic`. Right now this test doesn't actually do anything.

Comment (rupeshnemade) [reply tests/helm/common_test.go]: @swiatekm-sumo Updated to different value. Can you please approve PR checker run?

File: tests/helm/testdata/goldenfile/services_with_service_monitor_labels/services_with_service_monitor_labels/default.output.yaml [hunk 4]
```diff
@@ -24,191 +25,3 @@ spec:
       port: 8888
       targetPort: 8888
       protocol: TCP
----
```
[the remark this answers is not in the document]
Comment (rupeshnemade) [reply tests/helm/testdata/goldenfile/services_with_service_monitor_labels/services_with_service_monitor_labels/default.output.yaml]: @swiatekm-sumo Thanks for pointing this out. I have reverted output files.

File: tests/helm/testdata/goldenfile/logs_otc_daemonset/additional.output.yaml [hunk 5]
```diff
@@ -95,148 +97,15 @@ spec:
...
+          # yamllint disable-line rule:line-length
           image: public.ecr.aws/docker/library/busybox:1.36.0
           imagePullPolicy: IfNotPresent
           securityContext:
             capabilities:
-              drop:
-                - ALL
               add:
                 - CAP_CHOWN
-          command:
```
Comment (swiatekm-sumo) [inline tests/helm/testdata/goldenfile/logs_otc_daemonset/additional.output.yaml:106]: Why does this change at all?

File: tests/helm/testdata/goldenfile/logs_otc_daemonset/additional.input.yaml [hunk 6]
```diff
@@ -1,3 +1,4 @@
+namespaceOverride: "sumologic"
```
Comment (swiatekm-sumo) [inline tests/helm/testdata/goldenfile/logs_otc_daemonset/additional.input.yaml:1]: I think this shouldn't be here.

File: tests/helm/testdata/goldenfile/logs_otc_daemonset/additional.output.yaml [hunk 7]
```diff
@@ -95,148 +97,15 @@ spec:
               protocol: TCP
       initContainers: # ensure the host path is owned by the otel user group
         - name: changeowner
+          # yamllint disable-line rule:line-length
           image: public.ecr.aws/docker/library/busybox:1.36.0
           imagePullPolicy: IfNotPresent
           securityContext:
             capabilities:
-              drop:
-                - ALL
```
Comment (swiatekm-sumo) [inline tests/helm/testdata/goldenfile/logs_otc_daemonset/additional.output.yaml:103]: Why does anything other than the namespace change in this file?

File: tests/helm/testdata/goldenfile/opentelemetry_operator_instrumentation_cr_job/job.output.yaml [hunk 8]
```diff
@@ -1,66 +1,37 @@
 ---
-# Source: sumologic/templates/opentelemetry-operator/job-instrumentation-cr.yaml
```
Comment (swiatekm-sumo) [inline tests/helm/testdata/goldenfile/opentelemetry_operator_instrumentation_cr_job/job.output.yaml:2]: Why does anything change in this file?

Comment (rupeshnemade) [reply tests/helm/testdata/goldenfile/opentelemetry_operator_instrumentation_cr_job/job.output.yaml]: @swiatekm-sumo I have reverted unwanted changes added by local output file regenerate. Didn't noticed these automatic updates from local make regenerate.

File: tests/helm/testdata/goldenfile/logs_otc_daemonset/additional.output.yaml (hunk 7 above)
Comment (rupeshnemade) [reply tests/helm/testdata/goldenfile/logs_otc_daemonset/additional.output.yaml]: @swiatekm-sumo I have reverted unwanted changes added by local output file regenerate. Didn't noticed these automatic updates from local make regenerate.

File: tests/helm/testdata/goldenfile/logs_otc_daemonset/additional.input.yaml (hunk 6 above)
Comment (rupeshnemade) [reply tests/helm/testdata/goldenfile/logs_otc_daemonset/additional.input.yaml]: @swiatekm-sumo Removed.

File: deploy/helm/sumologic/templates/_helpers/_logs.tpl [hunk 9]
```diff
@@ -299,4 +299,4 @@ Return the otelcol log collector image
 */}}
 {{- define "sumologic.logs.collector.image" -}}
 {{ template "utils.getOtelImage" (dict "overrideImage" .Values.otellogs.image "defaultImage" .Values.sumologic.otelcolImage) }}
-{{- end -}}
+{{- end -}}
```
[the remark this answers is not in the document]
Comment (rupeshnemade) [reply deploy/helm/sumologic/templates/_helpers/_logs.tpl]: @swiatekm-sumo Added.

File: deploy/helm/sumologic/conf/events/fluentd/events.conf [hunk 10]
```diff
@@ -13,7 +13,7 @@
 {{- else }}
 <source>
   @type events
-  deploy_namespace {{ .Release.Namespace }}
+  deploy_namespace {{ template "sumologic.namespace" . }}
```
Comment (astencel-sumo) [inline deploy/helm/sumologic/conf/events/fluentd/events.conf:16]: What about the reference to `{{ $.Release.Namespace }}` in line 5 above?

**Comment by astencel-sumo** [comment]
Here's the result of searching for `.Release.Namespace` in the repo. I suppose all the leftover references should be replaced with `sumologic.namespace` template?

image

File: deploy/helm/sumologic/conf/events/fluentd/events.conf (hunk 10 above)
Comment (rupeshnemade) [reply deploy/helm/sumologic/conf/events/fluentd/events.conf]: @astencel-sumo Thanks for pointing that out. I have replaced all remaining occurrences of .Release.Namespace

**Comment by rupeshnemade** [comment]
> Here's the result of searching for `.Release.Namespace` in the repo. I suppose all the leftover references should be replaced with `sumologic.namespace` template?
> 
> image
@astencel-sumo Thanks for pointing that out. I have replaced all remaining occurrences of .Release.Namespace

**Review by swiatekm-sumo — APPROVED** [review]

File: deploy/helm/sumologic/templates/clusterrole.yaml [hunk 11]
```diff
@@ -2,6 +2,7 @@ apiVersion: rbac.authorization.k8s.io/v1
 kind: ClusterRole
 metadata:
   name: {{ template "sumologic.metadata.name.roles.clusterrole" . }}
+  namespace: {{ template "sumologic.namespace"  . }}
```
Comment (astencel-sumo) [inline deploy/helm/sumologic/templates/clusterrole.yaml:5]: `ClusterRole`s are not namespaced
```suggestion
```

File: deploy/helm/sumologic/templates/clusterrolebinding.yaml [hunk 12]
```diff
@@ -2,12 +2,13 @@ apiVersion: rbac.authorization.k8s.io/v1
 kind: ClusterRoleBinding
 metadata:
   name: {{ template "sumologic.metadata.name.roles.clusterrolebinding" . }}
+  namespace: {{ template "sumologic.namespace"  . }}
```
Comment (astencel-sumo) [inline deploy/helm/sumologic/templates/clusterrolebinding.yaml:5]: `ClusterRoleBinding`s are not namespaced
```suggestion
```

File: tests/helm/common_test.go [hunk 13]
```diff
@@ -242,3 +242,40 @@ func GetPodSpec(object unstructured.Unstructured) (*corev1.PodSpec, error) {
...
+	// split the rendered Yaml into individual documents and unmarshal them into K8s objects
+	// we could use the yaml decoder directly, but we'd have to implement our own unmarshaling logic then
+	renderedObjects := UnmarshalMultipleFromYaml[unstructured.Unstructured](t, renderedYamlString)
+
+	for _, renderedObject := range renderedObjects {
+		if !isSubchartObject(&renderedObject) && renderedObject.GetKind() != "List" {
+			object := renderedObject
+			objectName := fmt.Sprintf("%s/%s", object.GetKind(), object.GetName())
+			t.Run(objectName, func(t *testing.T) {
+				require.Equal(t, namespaceOverride, object.GetNamespace())
```
Comment (swiatekm-sumo) [inline tests/helm/common_test.go:277]: Given @astencel-sumo 's comment about ClusterRoles and ClusterRoleBindings, you'll need to change this to:

```suggestion
                                namespace := object.GetNamespace()
                                if namespace != "" {
				  require.Equal(t, namespaceOverride, object.GetNamespace())
				}
```

to account for non-namespaced resources.

**Comment by rupeshnemade** [comment]
@astencel-sumo After helm upgrade it looks like helm has picked 'sumo' namespace based on output-
`Check the release status by running:
  kubectl --namespace sumo get pods -l "release=col"`
Need to verify if its Helm issue.

For Prometheus we will need to add in docs to explicitly provide namespaceOverride in kube-prometheus-stack.
I will work on adding integration test for Prometheus scenario. 
@swiatekm-sumo Should existing TestNamespaceOverride test need to have this functionality or I need to add complete new test case?

**Comment by swiatekm-sumo** [comment]
> @swiatekm-sumo Should existing TestNamespaceOverride test need to have this functionality or I need to add complete new test case?

You need to add a new integration test in `tests/integration`. Might be a bit tricky, as the tests create a separate namespace for each test by default. You'd want to copy [the default test file](https://github.com/SumoLogic/sumologic-kubernetes-collection/blob/main/tests/integration/helm_ot_default_test.go), add a values file with the namespace overridden, and then change the namespace set in the context to the overridden one.

File: deploy/helm/sumologic/templates/clusterrolebinding.yaml (hunk 12 above)
Comment (rupeshnemade) [reply deploy/helm/sumologic/templates/clusterrolebinding.yaml]: @astencel-sumo This has been now removed. Earlier goldentest needed it so we added it here but ideally yes we don't need namsepace for non-namespaced resource.

File: deploy/helm/sumologic/templates/clusterrole.yaml (hunk 11 above)
Comment (rupeshnemade) [reply deploy/helm/sumologic/templates/clusterrole.yaml]: @astencel-sumo This has been now removed. Earlier goldentest needed it so we added it here but ideally yes we don't need namsepace for non-namespaced resource.

**Comment by swiatekm-sumo** [comment]
@rupeshnemade have you figured out the cause of `helm upgrade`'s behaviour with the namespace override?

**Review by astencel-sumo — APPROVED** [review]

**Pull request merged**